File: src/2025-02-23-deepclaude-try/javascript/exercises/practice/zipper/zipper.js
//
// This is only a SKELETON file for the 'Zipper' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Zipper {
  constructor(tree, breadcrumbs) {
    this._tree = tree;
    this._breadcrumbs = breadcrumbs;
  }

  static fromTree(tree) {
    return new Zipper(tree, []);
  }

  toTree() {
    let current = this;
    while (current._breadcrumbs.length > 0) {
      current = current.up();
    }
    return current._tree;
  }

  value() {
    return this._tree?.value ?? null;
  }

  left() {
    if (!this._tree?.left) return null;
    return new Zipper(this._tree.left, [
      ...this._breadcrumbs,
      {
        value: this._tree.value,
        right: this._tree.right,
        direction: 'left'
      }
    ]);
  }

  right() {
    if (!this._tree?.right) return null;
    return new Zipper(this._tree.right, [
      ...this._breadcrumbs,
      {
        value: this._tree.value,
        left: this._tree.left,
        direction: 'right'
      }
    ]);
  }

  up() {
    if (this._breadcrumbs.length === 0) return null;
    const crumb = this._breadcrumbs[this._breadcrumbs.length - 1];
    const newBreadcrumbs = this._breadcrumbs.slice(0, -1);
    
    const parentTree = {
      value: crumb.value,
      left: crumb.direction === 'left' ? this._tree : crumb.left,
      right: crumb.direction === 'right' ? this._tree : crumb.right
    };
    
    return new Zipper(parentTree, newBreadcrumbs);
  }

  setValue(value) {
    if (!this._tree) return this;
    const newTree = { ...this._tree, value };
    return new Zipper(newTree, this._breadcrumbs);
  }

  setLeft(left) {
    if (!this._tree) return this;
    const newTree = { ...this._tree, left };
    return new Zipper(newTree, this._breadcrumbs);
  }

  setRight(right) {
    if (!this._tree) return this;
    const newTree = { ...this._tree, right };
    return new Zipper(newTree, this._breadcrumbs);
  }
}

File: src/python-2025-02-23-yh-deepclaude-retry/python/exercises/practice/zipper/zipper.py
class Zipper:
    def __init__(self, focus, path):
        self.focus = focus
        self.path = path

    @staticmethod
    def from_tree(tree):
        return Zipper(tree, [])

    def value(self):
        return self.focus['value']

    def set_value(self, value):
        new_focus = {'value': value, 'left': self.focus['left'], 'right': self.focus['right']}
        return Zipper(new_focus, self.path)

    def left(self):
        if self.focus['left'] is None:
            return None
        new_focus = self.focus['left']
        new_path = [('left', self.focus['value'], self.focus['right'])] + self.path
        return Zipper(new_focus, new_path)

    def set_left(self, subtree):
        new_focus = {'value': self.focus['value'], 'left': subtree, 'right': self.focus['right']}
        return Zipper(new_focus, self.path)

    def right(self):
        if self.focus['right'] is None:
            return None
        new_focus = self.focus['right']
        new_path = [('right', self.focus['value'], self.focus['left'])] + self.path
        return Zipper(new_focus, new_path)

    def set_right(self, subtree):
        new_focus = {'value': self.focus['value'], 'left': self.focus['left'], 'right': subtree}
        return Zipper(new_focus, self.path)

    def up(self):
        if not self.path:
            return None
        direction, parent_val, sibling = self.path[0]
        if direction == 'left':
            new_focus = {'value': parent_val, 'left': self.focus, 'right': sibling}
        else:
            new_focus = {'value': parent_val, 'left': sibling, 'right': self.focus}
        return Zipper(new_focus, self.path[1:])

    def to_tree(self):
        z = self
        while z.path:
            z = z.up()
        return z.focus
